### PR TITLE
(v2) adaptive colors + writers

### PR DESCRIPTION
**Pull request by meowgorithm**
This PR contains solutions for adaptive colors, background color detection, and colorprofile-based writers. It also updates all tests and examples for `v2`.

The big changes in here are:

```go
// Background color detection.
hasDarkBG, err := lipgloss.HasDarkBackground(os.Stdin, os.Stdout)
if err != nil {
    log.Fatal("Uh oh:", err)
}

// Adaptive color helper.
lightDark := lipgloss.LightDark(hasDarkBG)
c := lightDark.Color(lightDark(0x212121, 0xf1f1f1))

// Color downsampling happens via a writer.
str := lipgloss.NewStyle().Foreground(c).Render("Hello!")
lipgloss.Println(str)
```

Available writer utilties are:

* `Writer` (writes to stdout)
* `Print(v ...any)`
* `Println(...any)`
* `Printf(string, ...any)`
* `Fprint(io.Writer, ...any)`
* `Fprintf(io.Writer, string, ...any)`

Supersedes #392.
Supersedes #389.

### CODE REVIEW

File: examples/layout/main.go [hunk 1]
```diff
@@ -309,16 +334,23 @@ func main() {
...
 	{
 		w := lipgloss.Width
 
+		lightDarkState := func() string {
+			if hasDarkBG {
+				return "Dark"
+			}
+			return "Light"
+		}()
+
```
Comment (aymanbagabas) [inline examples/layout/main.go:347]:
```suggestion
		lightDarkState := "Light"
		if hasDarkBG {
			lightDarkState = "Dark"
		}

```

File: go.mod [hunk 2]
```diff
@@ -8,16 +8,18 @@ go 1.18
...
 require (
 	github.com/aymanbagabas/go-udiff v0.2.0
-	github.com/charmbracelet/x/ansi v0.2.3
+	github.com/charmbracelet/colorprofile v0.1.2
+	github.com/charmbracelet/x/ansi v0.3.2
 	github.com/charmbracelet/x/exp/golden v0.0.0-20240806155701-69247e0abc2a
+	github.com/charmbracelet/x/input v0.2.0
+	github.com/charmbracelet/x/term v0.2.0
 	github.com/lucasb-eyer/go-colorful v1.2.0
-	github.com/muesli/termenv v0.15.2
```
Comment (aymanbagabas) [inline go.mod:14]: 👀

File: writer.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+// Example:
+//
+//	str := NewStyle().
+//	    Foreground(lipgloss.Color("#6a00ff")).
+//	    Render("artichokes")
+//
+//	Fprintf(os.Stderr, "I really love %s!\n", food)
+func Fprintf(w io.Writer, format string, v ...interface{}) (int, error) {
+	return fmt.Fprintf(colorprofile.NewWriter(w, os.Environ()), format, v...) //nolint:wrapcheck
+}
```
Comment (aymanbagabas) [inline writer.go:96]: I would leave this out. Every call allocates a new writer which defeats the purpose of `fmt.Fprint`. Instead, people should create a writer for `os.Stderr` and use `fmt.Fprint`

```go
w := colorprofile.NewWriter(os.Stderr, os.Environ())
fmt.Fprintf(w, "I really love %s!", lipgloss.NewStyle().Render("artichokes"))
```

File: examples/layout/main.go (hunk 1 above)
Comment (meowgorithm) [reply examples/layout/main.go:347]: Farewell, my λ

File: writer.go (hunk 3 above)
Comment (meowgorithm) [reply writer.go:96]: I agree, but ideally we reduce the cognitive complexity a bit. Let's look at it again once this beast of a branch is merged into `v2-exp`.

**Pull request merged**